code/xcvm/cosmwasm/contracts/gateway/src/contract/ibc/ics20.rs

```diff
@@ -35,68 +39,137 @@ pub(crate) fn handle_bridge_forward(
+	let (msg, event) = if let Ok(transfer_shortcut) =
+		ibc_ics_20_transfer_shortcut(deps.as_ref(), &msg)
+	{
+		let mut _event = make_event("bridge")
+			.add_attribute("to_network_id", msg.to.to_string())
+			.add_attribute("shortcut", "ics20-transfer");
+
+		unimplemented!("add tracking lock for funds return usual cosmos message to transfer as defined in {:?}", transfer_shortcut);
+	} else {
+		let route: IbcIcs20ProgramRoute = get_this_route(deps.storage, msg.to, *local_asset)?;
+		state::tracking::bridge_lock(deps.storage, (msg.clone(), route.clone()))?;
+
+		let asset = msg
+			.msg
+			.assets
+			.0
+			.get(0)
+			.map(|(_, amount)| (route.on_remote_asset, *amount))
+			.expect("not empty");
+
+		let packet = XcPacket {
+			interpreter: String::from(info.sender).into_bytes(),
+			user_origin: msg.interpreter_origin.user_origin.clone(),
+			salt: msg.msg.salt,
+			program: msg.msg.program,
+			assets: vec![asset].into(),
+		};
+
+		deps.api.debug(&format!(
+			"cvm::gateway::ibc::ics20 route {}",
+			&serde_json_wasm::to_string(&route)?
+		));
+
+		let mut event = make_event("bridge")
+			.add_attribute("to_network_id", msg.to.to_string())
+			.add_attribute(
+				"assets",
+				serde_json_wasm::to_string(&packet.assets)
+					.map_err(|_| ContractError::FailedToSerialize)?,
+			)
+			.add_attribute(
+				"program",
+				serde_json_wasm::to_string(&packet.program)
+					.map_err(|_| ContractError::FailedToSerialize)?,
+			);
+
+		if !packet.salt.is_empty() {
+			let salt_attr = Binary::from(packet.salt.as_slice()).to_string();
+			event = event.add_attribute("salt", salt_attr);
+		}
 
-	let msg = to_cw_message(deps.as_ref(), deps.api, coin, route, packet, block)?;
+		let coin = Coin::new(amount.0, route.local_native_denom.clone());
+
+		match route.gateway_to_send_to.clone() {
+			GatewayId::CosmWasm { contract, .. } => {
+				let msg = to_cosmwasm_message(
+					deps.as_ref(),
+					deps.api,
+					coin,
+					route,
+					packet,
+					block,
+					contract,
+				)?;
+				(msg, event)
+			},
+			GatewayId::Evm { .. } => Err(ContractError::NotImplemented)?,
+		}
+	};
 
 	Ok(Response::default()
 		.add_event(event)
 		.add_submessage(SubMsg::reply_on_success(msg, ReplyId::TransportSent.into())))
 }
 
+/// When target network supports native cross chain operation of Transfer,
+/// and program as simple as just Transfer,
+/// can use instance of this structure to route pure funds transfer
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+#[cfg_attr(feature = "std", derive(schemars::JsonSchema))]
+#[serde(rename_all = "snake_case")]
+pub struct IbcIcs20TransferShortcutRoute {
+	pub source: ChannelId,
+	pub denom: String,
+	pub sending: xc_core::transport::ibc::IbcIcs20Sender,
+}
+
+/// this method return route in case program can be just transfer
+pub fn ibc_ics_20_transfer_shortcut(
+	deps: Deps,
+	msg: &xc_core::gateway::BridgeForwardMsg,
+) -> Result<IbcIcs20TransferShortcutRoute, ContractError> {
```

Comment:
This sounds like `-> Result<Option<IbcIcs20TransferShortcutRoute>, ContractError>` to me and `Ok(None)` returned if use_shortcut is false or unset I guess.

Reply:
option with/vs result opinionated. 

my opinion if there is result there is no need option, given rust made results as ergonomic as options, do not see reason to use option. 

also soon rust will have keyword generics around result (at least they say), even more less need to use option.

also in case of error like no asset/connection - neither option/result observable as it tries to invoke general program, so it does not improves debugging.

also some languages ditch options, like Roc, as lead to more onsistentencies, Option<T>, is just Result<T,()>

Reply:
If the error this function returns doesn’t matter and is never inspected than this should be `Option<IbcIcs20TransferShortcutRoute>` then.  If the function returns an error and the error is ignored, that raises questions. My understanding of this function is that it can return:
- an error which should be propagated,
- `None` if shortcut is disabled or
- `Some(shortcut)` if shortcut is enabled.

Thus my suggestion is `Result<Option<_>, _>`. This isn’t opinionated. It better documents what the function returns.

Reply:
> if the function returns an error and the error is ignored, that raises questions. 

why than not raises question ignoring errors inside function if it returns Option? and error is not ignored, it just tells that there is no shortcut for some reason. later I can log it just to see traces from production on routes resolutions.  so i better return error and have logs higher level, than make option and prevent higher level to debug.

Reply:
> an error which should be propagated,
> None if shortcut is disabled or
> Some(shortcut) if shortcut is enabled.
> Thus my suggestion is Result<Option<_>, _>. This isn’t opinionated. It better documents what the function returns.

I understand these words, but thing is i do not see how it helps ergonomics/debugging, it may help some readers, but that is opinionated.

Reply:
If I would return Option<T>, but one of methods called inside returns error. Would `?` make Result in case of error become None? So I do not have to call something `.map_error_to_option()`.

Reply:
I would be happy to document failing to solve shortcut as one of enum variants.

Method tries to get shortcut route, it cannot get, it errors. Why? Because either no connection or shortcut disabled, or other reason. For me option is just another case. Specifically direct IBC ICS20 connection not found. But later for sure will have non direct solutions by unrolling prefix path (need to build index for that so). In this case no direct connection means nothing. So would no direction be error or option? I would prefer error so I can log it for tracing.

Reply:
> If I would return Option, but one of methods called inside returns error. Would ? make Result in case of error become None? So I do not have to call something .map_error_to_option().

If one of the calls return an error, why isn’t it propagated? (And for function returning Option, dealing with Result is just `.ok()?`).

> Method tries to get shortcut route, it cannot get, it errors. Why? Because either no connection or shortcut disabled, or other reason. 

But shortcut being disabled is not an error. Unless I misunderstand something, there are three possible options: shortcut, no shortcut or error. If shortcut is enabled and resolved correctly, caller uses the shortcut; if shortcut is disabled, caller uses full path; and if some error while figuring out error happens, caller propagates that error.  This is best modelled by `Result<Option<_>, _>`.  With plain `Result<_>` the caller must now match on the error type to decide whether error needs to be propagated or ‘no shortcut’ path chosen.